### ISSUE TITLE
How to track request or response headers not currently tracked ?

### ISSUE THREAD

**Issue by valeriob** [microsoft/ApplicationInsights-Home#406]
I would like to track more headers from both requests and responses, for example the presence of some Auth header on the request or the size of the content on the response.
What would the correct api be ?

Thanks
Valerio

**Comment by Dmitry-Matveev** [comment]
You can give a shot to [Telemetry Initializers](https://docs.microsoft.com/en-us/azure/azure-monitor/app/api-filtering-sampling). They are invoked synchronously to telemetry item creation, so for Request Telemetry and Dependency Telemetry they will be in the context of the currently executed call.

**Comment by valeriob** [comment]
Hi @Dmitry-Matveev i tried that, but i can not reach the HttpContext to extract the information.

**Comment by Dmitry-Matveev** [comment]
I guess, as a workaround, you can store Headers into AsyncLocal at the moment you can access them and then, in Telemetry Initializer read them from AsyncLocal. However, that means you'd need to get to headers before Initializers are invoked, so you'd do it before the incmoing call is complete.

**Comment by valeriob** [comment]
Thanks Dmitry, i'm not a fan of using AsyncLocal, i think that accessing the http context should be a main feature of the library, anyway i'll give it a try, when do you suggest i should store the data  to be able to retrieve it from the Telemetry Initializer ?
Thanks

**Comment by valeriob** [comment]
Hi, i tryed to use asyncLocal without luck, i'm intercepting it in a middlewhare but when the code inside the ITelemetryInitializer runs, the value is null.
Can you provide some other way? It's very important to add some context to the httprequest to diagnose prdocution issues.
Thanks

**Comment by Dmitry-Matveev** [comment]
I thought of something like this:
```
    class MyInitializer : ITelemetryInitializer
    {
        public static AsyncLocal<string> MyAsyncStorage { get; } = new AsyncLocal<string>();

        public void Initialize(ITelemetry telemetry)
        {
            if (telemetry is RequestTelemetry request)
            {
                if (MyAsyncStorage.Value != null)
                {
                    request.NecessaryFieldHere = MyAsyncStorage.Value;
                }
            }
        }
    }
```
Then, for incoming request, the moment when you first can access your headers, save necessary value to the MyAsyncStorage.

I may simply be missing something and there can be a better way for getting headers to initializers, @lmolkova is there a better suggestion?

**Comment by vborioni-onit** [comment]
Yep i tried that, then i set the storage value in a middleware : 

            app.Use(async (context, next) =>
            {
                MyTelemetryInitializer.AsyncLocal.Value = context;
                await next.Invoke();
            });

**Comment by Dmitry-Matveev** [comment]
@valeriob , @vborioni-onit , (not sure which alias is current), I consulted with @lmolkova and she suggested to try storing the values you'd like to use in `HttpContext.Current.Items` and then taking them from there in Telemetry Initializer. I tried it on a sample Web API app and it works.

Alternatively, you can save them directly to `RequestTelemetry` object that is stored in `HttpContext.Current.Items` as soon as `RequestTelemetry` is created. However, I cannot recommend this as a first choice as it's not public API (you'll be relying on the name of our variable in `HttpContext.Current.Items`) and you do not control the moment when AI adds `RequestTelemetry` to that collection - so it may be empty at the moment you'd like to modify it.

**Comment by valeriob** [comment]
Hi @Dmitry-Matveev the samples refer to aspnet MVC 5, i'm talking about aspnet core, there is no HttpContext.Current api in that framework.
